### PR TITLE
Abmarl 296 open spiel adapter

### LINKED ISSUE #296: Adapter for openspiel
see this comment below: https://github.com/LLNL/Abmarl/issues/296#issuecomment-1189517950

### PR DESCRIPTION
**Pull request by rusu24edward**
Added a wrapper for Agent Based Simulations to work with OpenSpiel rl algorithms.

Resolves #296

### CODE REVIEW

**Comment by rusu24edward** [comment]
Add `OpenSpielWrapper` to the `__init__.py` file.

File: abmarl/external/open_spiel_env_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+from gym.spaces import Discrete
+from open_spiel.python.rl_environment import TimeStep, StepType
+
+from abmarl.sim.agent_based_simulation import Agent
+from abmarl.managers import TurnBasedManager, SimulationManager
+
+
+class OpenSpielWrapper:
+    """
+    Enable connection between SimulationManager and OpenSpiel agents.
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:11]: Abmarl's Simulation Manager

File: abmarl/external/open_spiel_env_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+    OpenSpiel support turn-based and simultaneous simulations, which Abmarl provides
+    through the TurnBasedManager and AllStepManager. OpenSpiel expects TimeStep
+    objects as output, which include the observations, rewards, and step type.
+    Among the observations, it expects a list of legal actions availbe to the agent.
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:16]: availbe -> available

File: abmarl/external/open_spiel_env_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    The OpenSpielWrapper converts output from the simulation manager to the expected
+    format. A TimeStep output typically looks like this:
+        TimeStpe(
+            observations={
+                info_state: {agent_id: agent_obs for agent_id in agents},
+                legal_actions: {agent_id: agent_legal_actions for agent_id in agents},
+                current_player: current_agent_id
+            }
+            rewards={
+                {agnet_id: agent_reward for agnt_id in agents}
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:26]: agnet -> agent

Comment (rusu24edward) [reply abmarl/external/open_spiel_env_wrapper.py:26]: agnt_id -> agent_id

File: abmarl/external/open_spiel_env_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+            }
+            discounts={
+                {agent_id: agent_discout for agent_id in agents}
+            }
+            step_type=StepType enum
+        )
+
+    Furthermore, OpenSpiel provides actions as a list. This wrapper converts it
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:34]: This wrapper -> The OpenSpielWrapper
delete it

File: abmarl/external/open_spiel_env_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    those actions to a dict before forwarding it to the underlying simulation manager.
+
+    OpenSpiel does not support the ability for some agents of a game to finish
+    before others. The game is either ongoing, in which all agents are providing
+    actions, or else it is done for all agents. In contrast, Abmarl allows some agents to be
+    done before others while the game is still going. Abmarl expects that done
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:40]: while the game... -> as the simulation progresses.

File: abmarl/external/open_spiel_env_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    agents will not provide actions. OpenSpiel, however, will always provide actions
+    for all players. So this wrapper removes the actions from agents that are
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:42]: this wrapper -> The OpenSpielWrapper

File: abmarl/external/open_spiel_env_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    already done before forwarding the action to the underlying simulation manager.
+    Furthermore, OpenSpiel expects every agent to be present in the TimeStep outputs.
+    Normally, Abmarl will not provide output for agents that are done since they
+    have finished generating data in this episode. In order to work with OpenSpiel,
+    this wrapper forces output from all agents at every step.
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:47]: The OpenSpielWrapper

Comment (rusu24edward) [reply abmarl/external/open_spiel_env_wrapper.py:47]: every step, including those already done.

File: abmarl/external/open_spiel_env_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+            }
+        else: # type(value) == dict
+            for discount_id, discount in value.items():
+                assert discount_id in self._learning_agents, \
+                    "id for the discount must be an agent id."
+                assert type(discount) in (float, int), \
+                    "discount values must be a number."
+            assert all([
+                True if agent_id in value.keys() else False for agent_id in self._learning_agents
+            ]), "All agents must be given a discounted value."
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:100]: discount, not discounted value

File: abmarl/external/open_spiel_env_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+    def num_players(self):
+        """
+        The number of player is the number of learning agents in the simulation.
+        """
+        return sum([1 for _ in self._learning_agents])
+
+    @property
+    def is_turn_based(self):
+        """
+        The game is turn based if the simulation is wrapped with a TurnBasedManager.
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:113]: if the simulation is a TurnBasedManager.

File: abmarl/external/open_spiel_env_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+
+    def reset(self, **kwargs):
+        """
+        Reset the simulation.
+        """
+        self._should_reset = False
+        obs = self.sim.reset(**kwargs)
+        self.current_player = next(iter(obs))
+
+        # If it is a turn based sim, then the current player should be the next
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:143]: Remove this comment since we figured it out

File: abmarl/external/open_spiel_env_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+        return self._current_player
+
+    @current_player.setter
+    def current_player(self, value):
+        assert value in self._learning_agents, "Current player must be an agent in the simulation."
+        self._current_player = value
+
+    def reset(self, **kwargs):
+        """
+        Reset the simulation.
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:137]: Return

File: abmarl/external/open_spiel_env_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+        return TimeStep(
+            observations=observations,
+            rewards=None,
+            discounts=None,
+            step_type=StepType.FIRST
+        )
+
+    def step(self, action_list, **kwargs):
+        """
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:162]: Args and returns

File: abmarl/external/open_spiel_env_wrapper.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+        Step the simulation forward using the reported actions.
+
+        OpenSpiel provides an action list of either (1) the agent whose turn it
+        is in a turn-based game or (2) all the agents in a simultaneous game. The
+        OpenSpielWrapper converts the list of actions to a dictionary before passing
+        it to the underlying simulation.
+
+        OpenSpiel does not support the ability for some agents of a game to finish
+        before others. As such, it may provide actions for agents that are already
+        done. To work with Abmarl, this wrapper removes actions for agents that
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:172]: "this wrapper"

File: abmarl/external/open_spiel_env_wrapper.py [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+        are already done.
+        """
+        # Actions come in as a list, so we need to convert to a dict before forwarding
+        # to the SimulationManager.
+        if self.is_turn_based:
+            action_dict = {self.current_player: action_list[0]}
+        else:
+            action_dict = {
+                agent_id: action_list[i]
+                for i, agent_id in enumerate(self._learning_agents)
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:182]: add assertion for simultaneous simulation that the action list is the same length as the learning agents.

File: abmarl/external/open_spiel_env_wrapper.py [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+            }
+        # OpenSpiel can send actions for agents that are already done, which doesn't
+        # work with our simulation managers. So we filter out these actions before
+        # passing them to the manager.
+        # TODO: Although all implemeted managers do track the done agents, this is not
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:187]: Need to do this!

File: abmarl/external/open_spiel_env_wrapper.py [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+        # a part of the SimulationManager interface.
+        for agent_id in self.sim.done_agents:
+            try:
+                del action_dict[agent_id]
+            except KeyError:
+                pass
+        # We have just deleted actions from agents that are already done, which
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:194]: from -> for

File: abmarl/external/open_spiel_env_wrapper.py [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+        # can result in an empty action dictionary (e.g. in a turn-based game).
+        # In this case, we just take a fake step.
+        if not action_dict: # No actions
+            return self._take_fake_step()
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:198]: Add a test in the test suite with an empty dictionary to check that fake stepping works

File: abmarl/external/open_spiel_env_wrapper.py [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+
+        if self._should_reset:
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:200]: Move this to top of step function

File: abmarl/external/open_spiel_env_wrapper.py [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+        return {
+            agent.id: {
+                'num_actions': agent.action_space.n,
+                'min': 0,
+                'max': agent.action_space.n - 1,
+                'dtype': int
+            } for agent in self._learning_agents.values()
+        }
+
+    def get_legal_actions(self, agent_id):
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:256]: Consider modifying legal actions if the agent is done.

File: abmarl/external/open_spiel_env_wrapper.py [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+        """
+        Return the legal actions available to the agent.
+
+        By default, this wrapper uses all the available actions as the legal actions
+        in each time step. This function can be overwritten in a derived class
+        to add logic for obtaining the actual legal actions available.
+        """
+        return [i for i in range(self._learning_agents[agent_id].action_space.n)]
+
+    def _append_obs(self, obs):
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:266]: Consider using null obs for agents that are done

File: abmarl/external/open_spiel_env_wrapper.py [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+            rewards=self._append_reward(reward),
+            discounts=self._discounts,
+            step_type=step_type
+        )
+
+    def observation_spec(self):
+        """
+        The agents' observations spaces.
+
+        Abmarl uses gym spaces for the observation space. This wrapper converts
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:229]: this wrapper

**Review by rusu24edward — COMMENTED** [review]
- [x] Correct use of game with simulation.
- [x] Correct use of "player" with agent.
- [ ] Correct use of "this wrapper" with the OpenSpielWrapper

File: tests/external/test_open_spiel_wrapper.py [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+    assert time_step.discounts == sim._discounts
+    assert time_step.step_type == StepType.MID
+
+
+def test_take_fake_step():
+    sim = OpenSpielWrapper(
+        AllStepManager(abs)
+    )
+    sim.reset()
+    time_step = sim._take_fake_step()
```
Comment (rusu24edward) [inline tests/external/test_open_spiel_wrapper.py:171]: Change this to step with no action

File: abmarl/external/open_spiel_env_wrapper.py (hunk 18 above)
Comment (rusu24edward) [reply abmarl/external/open_spiel_env_wrapper.py:200]: Add test in test suite for should reset

File: abmarl/external/open_spiel_env_wrapper.py (hunk 20 above)
Comment (rusu24edward) [reply abmarl/external/open_spiel_env_wrapper.py:266]: Not gonna do for now because bug #304

File: abmarl/external/__init__.py [hunk 23]
```diff
@@ -1,2 +1,3 @@
 from .gym_env_wrapper import GymWrapper
 from .rllib_multiagentenv_wrapper import MultiAgentWrapper
+from .open_spiel_env_wrapper import OpenSpielWrapper
```
Comment (rusu24edward) [inline abmarl/external/__init__.py:3]: Extra line

File: abmarl/external/open_spiel_env_wrapper.py [hunk 24]
```diff
@@ -97,33 +98,33 @@ def discounts(self, value):
...
         """
-        The current player is the player that currently provides the action.
+        The agent that currently provides the action.
 
-        This is used the in the observation part of the TimeStep output. If it
-        is a turn based game, then the current player is the single agent who
-        is providing an action. If it is a simultaneous game, then OpenSpiel does
-        not use this property and the current player is just the first player
-        in the list.
+        Current player is used the in the observation part of the TimeStep output. If it
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:123]: remove the

File: abmarl/external/open_spiel_env_wrapper.py [hunk 25]
```diff
@@ -163,43 +163,53 @@ def step(self, action_list, **kwargs):
...
+        if self._should_reset:
+            return self.reset(**kwargs)
+
         # Actions come in as a list, so we need to convert to a dict before forwarding
         # to the SimulationManager.
         if self.is_turn_based:
             action_dict = {self.current_player: action_list[0]}
         else:
+            assert len(action_list) == len(self._learning_agents), \
+                "The number of actions must match the number of learning agnets " + \
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:193]: agnets -> agents

File: abmarl/external/open_spiel_env_wrapper.py [hunk 26]
```diff
@@ -241,7 +248,7 @@ def action_spec(self):
         """
         The agents' action spaces.
 
-        Abmarl uses gym spaces for the action space. This wrapper converts
+        Abmarl uses gym spaces for the action space. The OpenSpiel wrapper converts
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:251]: OpenSpielWrapper

File: abmarl/external/open_spiel_env_wrapper.py [hunk 27]
```diff
@@ -257,11 +264,11 @@ def get_legal_actions(self, agent_id):
         """
         Return the legal actions available to the agent.
 
-        By default, this wrapper uses all the available actions as the legal actions
-        in each time step. This function can be overwritten in a derived class
+        By default, the OpenSpielWrapper wrapper uses the agent's entire action space
```
Comment (rusu24edward) [inline abmarl/external/open_spiel_env_wrapper.py:267]: delete extra "wrapper"
